pkg/controllers/tagging/tagging_controller.go

```diff
@@ -0,0 +1,289 @@
+				// Put the item back on the workqueue to handle any transient errors.
+				tc.workqueue.AddRateLimited(key)
+				return fmt.Errorf("error tagging '%s': %s, requeuing", key, err.Error())
+			}
+		} else {
+			if err := tc.untagNodeResources(node); err != nil {
+				tc.workqueue.AddRateLimited(key)
+				return fmt.Errorf("error untagging '%s': %s, requeuing", key, err.Error())
+			}
+		}
+
+		tc.workqueue.Forget(obj)
+		klog.Infof("Finished processing %v", obj)
+		return nil
+	}(obj)
+
+	if err != nil {
+		klog.Errorf("Error occurred while processing %v", obj)
+		utilruntime.HandleError(err)
+	}
+
+	return true
+}
+
+// tagNodesResources tag node resources from a list of nodes
+// If we want to tag more resources, modify this function appropriately
+func (tc *TaggingController) tagNodesResources(node *v1.Node) error {
+	for _, resource := range tc.resources {
+		switch resource {
+		case opt.Instance:
+			err := tc.tagEc2Instance(node)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// tagEc2Instances applies the provided tags to each EC2 instance in
+// the cluster.
+func (tc *TaggingController) tagEc2Instance(node *v1.Node) error {
+	instanceId, err := awsv1.KubernetesInstanceID(node.Spec.ProviderID).MapToAWSInstanceID()
+
+	if err != nil {
+		klog.Errorf("Error in getting instanceID for node %s, error: %v", node.GetName(), err)
+		return err
+	} else {
+		err := tc.cloud.TagResource(string(instanceId), tc.tags)
+
+		if err != nil {
+			klog.Errorf("Error in tagging EC2 instance for node %s, error: %v", node.GetName(), err)
+			return err
+		}
+	}
+
+	klog.Infof("Successfully tagged %s with %v", instanceId, tc.tags)
+
+	return nil
+}
+
+// untagNodeResources untag node resources from a list of nodes
+// If we want to untag more resources, modify this function appropriately
+func (tc *TaggingController) untagNodeResources(node *v1.Node) error {
+	for _, resource := range tc.resources {
+		switch resource {
+		case opt.Instance:
+			err := tc.untagEc2Instance(node)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// untagEc2Instances deletes the provided tags to each EC2 instances in
+// the cluster.
+func (tc *TaggingController) untagEc2Instance(node *v1.Node) error {
+	instanceId, err := awsv1.KubernetesInstanceID(node.Spec.ProviderID).MapToAWSInstanceID()
+
+	if err != nil {
+		klog.Errorf("Error in getting instanceID for node %s, error: %v", node.GetName(), err)
+		return err
+	} else {
+		err := tc.cloud.UntagResource(string(instanceId), tc.tags)
+
+		if err != nil {
+			klog.Errorf("Error in untagging EC2 instance for node %s, error: %v", node.GetName(), err)
+			return err
+		}
+	}
+
+	klog.Infof("Successfully untagged %s with %v", instanceId, tc.tags)
+
+	return nil
+}
+
+// enqueueNode takes in the object to enqueue to the workqueue and whether
+// the object is to be tagged
+func (tc *TaggingController) enqueueNode(obj interface{}, toBeTagged bool) {
```

Comment:
Instead of passing in `toBeTagged` here which adds complexity here, can we take in a function as param which would basically be the action to be taken? For example, in case of `Add` or `Update` it could be `tagNodesResources(...)` and in case of `Delete` it can be `untagNodeResources(...)` and that way as part of processing the item, we just call the function/action associated with it?

Reply:
Could you please elaborate on this? As I understand, the change would render using the workqueue (and rate limiting) obsolete